pkg/kv/kvclient/rangefeed/db_adapter_external_test.go

```diff
@@ -53,35 +72,91 @@ func TestDBClientScan(t *testing.T) {
+				responses = append(responses, value)
+			}))
 		require.Len(t, responses, 0)
-	}
+	})
 
 	// Ensure that expected values are seen at the intermediate timestamp.
-	{
+	t.Run("scan sees values at intermediate ts", func(t *testing.T) {
 		var responses []roachpb.KeyValue
-		require.NoError(t, dba.Scan(ctx, sp, afterB, func(value roachpb.KeyValue) {
-			responses = append(responses, value)
-		}))
+		require.NoError(t, dba.ScanWithOptions(
+			ctx, []roachpb.Span{sp}, afterB, func(value roachpb.KeyValue) {
+				responses = append(responses, value)
+			}))
 		require.Len(t, responses, 2)
 		require.Equal(t, mkKey("a"), responses[0].Key)
 		va, err := responses[0].Value.GetInt()
 		require.NoError(t, err)
 		require.Equal(t, int64(1), va)
-	}
+	})
 
 	// Ensure that pagination doesn't break anything.
-	dba.SetTargetScanBytes(1)
-	{
+	t.Run("scan pagination works", func(t *testing.T) {
 		var responses []roachpb.KeyValue
-		require.NoError(t, dba.Scan(ctx, sp, db.Clock().Now(), func(value roachpb.KeyValue) {
-			responses = append(responses, value)
-		}))
+		require.NoError(t, dba.ScanWithOptions(ctx, []roachpb.Span{sp}, db.Clock().Now(),
+			func(value roachpb.KeyValue) {
+				responses = append(responses, value)
+			},
+			rangefeed.WithTargetScanBytes(1)))
 		require.Len(t, responses, 3)
-	}
+	})
+
+	// Ensure scan respects memory limits.
+	t.Run("scan respects memory limits", func(t *testing.T) {
+		const memLimit = 4096
+		mm := startMonitorWithBudget(memLimit)
+		defer mm.Stop(ctx)
+
+		require.Regexp(t, "memory budget exceeded",
+			dba.ScanWithOptions(ctx, []roachpb.Span{sp}, db.Clock().Now(),
+				func(value roachpb.KeyValue) {},
+				rangefeed.WithTargetScanBytes(2*memLimit),
+				rangefeed.WithMemoryMonitor(mm),
+			))
+	})
+
+	// Verify parallel scan operations.
+	t.Run("parallel scan requests", func(t *testing.T) {
+		sqlDB := sqlutils.MakeSQLRunner(tc.ServerConn(0))
+		sqlDB.Exec(t, `CREATE TABLE foo (key INT PRIMARY KEY)`)
+		sqlDB.Exec(t, `INSERT INTO foo (key) SELECT * FROM generate_series(1, 1000)`)
+		sqlDB.Exec(t, "ALTER TABLE foo SPLIT AT VALUES (250), (500), (750)")
+
+		fooDesc := catalogkv.TestingGetTableDescriptor(
+			db, keys.SystemSQLCodec, "defaultdb", "foo")
+		fooSpan := fooDesc.PrimaryIndexSpan(keys.SystemSQLCodec)
+
+		// We expect 4 splits -- we'll start the scan with parallelism set to 3.
+		// We will block these scans from completion until we know that we have 3
+		// concurrently running scan requests.
+		var parallelism = 3
+		var barrier int32 = 0
+		proceed := make(chan struct{})
+
+		g := ctxgroup.WithContext(context.Background())
+		g.GoCtx(func(ctx context.Context) error {
+			return dba.ScanWithOptions(ctx, []roachpb.Span{fooSpan}, db.Clock().Now(),
+				func(value roachpb.KeyValue) {},
+				rangefeed.WithInitialScanParallelismFn(func() int { return parallelism }),
+				rangefeed.WithOnSpanScanCompleted(func(ctx context.Context, sp roachpb.Span) {
```

Comment:
Can we just call this `OnScanCompleted` instead of `OnSpanScanCompleted`. The API makes it clear that it's per span.